x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/lock_manager_client.ts

```diff
@@ -0,0 +1,346 @@
+  }: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    return pRetry(async () => {
+      const acquired = await this.acquire({ metadata, ttl });
+      if (!acquired) {
+        this.logger.debug(`Lock "${this.lockId}" not available yet.`);
+        throw new Error(`Lock "${this.lockId}" not available yet`);
+      }
+      return acquired;
+    }, retryOptions ?? { forever: true, maxTimeout: 10_000 });
+  }
+
+  public async extendTtl(ttl = 300000): Promise<boolean> {
+    try {
+      await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        script: {
+          lang: 'painless',
+          source: `
+          if (ctx._source.token == params.token) {
+            long now = System.currentTimeMillis();
+            ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();
+          } else {
+            ctx.op = 'noop';
+          }`,
+          params: {
+            ttl,
+            token: this.token,
+          },
+        },
+        refresh: true,
+      });
+      this.logger.debug(`Lock "${this.lockId}" extended ttl with ${prettyMilliseconds(ttl)}.`);
+      return true;
+    } catch (error) {
+      if (isVersionConflictException(error)) {
+        this.logger.debug(`Lock "${this.lockId}" was released concurrently. Not extending TTL.`);
+        return false;
+      }
+
+      this.logger.error(`Failed to extend lock "${this.lockId}": ${error.message}`);
+      this.logger.debug(error);
+      return false;
+    }
+  }
+}
+
+export async function withLock<T>(
+  {
+    esClient,
+    logger,
+    lockId,
+    metadata,
+    ttl = duration(5, 'minutes').asMilliseconds(),
+    waitForLock = false,
+    retryOptions,
+  }: {
+    esClient: ElasticsearchClient;
+    logger: Logger;
+    lockId: LockId;
+    waitForLock?: boolean;
+    retryOptions?: pRetry.Options;
+  } & AcquireOptions,
+  callback: () => Promise<T>
+): Promise<T | undefined> {
+  const lockManager = new LockManager(lockId, esClient, logger);
+  const acquired =
+    waitForLock ?? retryOptions
+      ? await lockManager.acquireWithRetry({ metadata, ttl, ...retryOptions })
+      : await lockManager.acquire({ metadata, ttl });
+
+  if (!acquired) {
+    logger.debug(`Lock "${lockId}" not acquired. Exiting.`);
+    throw new Error(`Lock "${lockId}" not acquired`);
+  }
+
+  // extend the ttl periodically
+  const extendInterval = Math.floor(ttl / 2);
+  logger.debug(
+    `Lock "${lockId}" acquired. Extending TTL every ${prettyMilliseconds(extendInterval)}`
+  );
+  const extendPromises: Array<Promise<boolean>> = [];
+  const intervalId = setInterval(() => {
+    const p = lockManager.extendTtl().catch((err) => {
+      logger.error(`Failed to extend lock "${lockId}":`, err);
+      return false;
+    });
+    extendPromises.push(p);
+  }, extendInterval);
+
+  try {
+    return await callback();
+  } finally {
+    clearInterval(intervalId);
+    await lockManager.release();
```

Comment:
should we catch this one?

Reply:
we should already handle this in `release` but better safe than sorry. I'll catch